File: day18/main.py
import sys

filepath = sys.argv[1]
SIZE = int(sys.argv[2])
BYTES = int(sys.argv[3]) if len(sys.argv) > 3 else None


start = (SIZE-1, SIZE-1)
end = (0, 0)

with open(filepath, "r") as f:
    corruption = [
        tuple(map(int, s.strip().split(","))) for s in f.readlines()
    ]

def get_closest(unvisited, dist):
    return min(unvisited, key=lambda n: dist.get(n, float('inf')))

def get_unvisited_neighbors(x, y, unvisited):
    for dx, dy in ((-1, 0), (1, 0), (0, -1), (0, 1)):
        blah = x+dx, y+dy
        if blah in unvisited and x >= 0 and x < SIZE and y >= 0 and y < SIZE:
            yield blah

def try_find_path(n):
    dist = {start: 0}
    unvisited = set()
    for y in range(SIZE):
        for x in range(SIZE):
            if (x, y) not in corruption[:n]:
                unvisited.add((x, y))

    while unvisited:
        current = get_closest(unvisited, dist)
        for considered in get_unvisited_neighbors(*current, unvisited):
            try:
                new_cost = dist[current] + 1
            except KeyError:
                # path blocked
                return None
            if considered not in dist or new_cost < dist[considered]:
                dist[considered] = new_cost
        unvisited.remove(current)
        if end in dist:
            break
    return dist[end]

if BYTES:
    print("Part 1:", try_find_path(BYTES))

def find_least_satisfying(f, minimum, maximum):
    if f(minimum):
        return minimum
    high_false = minimum
    if not f(maximum):
        return None
    low_true = maximum
    while high_false < low_true-1:
        trial = (low_true - high_false)//2 + high_false
        if f(trial):
            low_true = trial
        else:
            high_false = trial
    return low_true

fewest_to_block = find_least_satisfying(
    lambda n: try_find_path(n) is None,
    BYTES or 0,
    len(corruption)-1,
)
print("Part 2:", ",".join(map(str, corruption[fewest_to_block-1])))

File: day18/main.rs
use std::collections::{HashMap, HashSet};
use std::fs::File;
use std::io::{self, BufRead};

fn main() -> io::Result<()> {
    let filepath = "input";
    let edge = 70;
    let (start, end) = ((0, 0), (edge, edge));

    let f = File::open(filepath)?;
    let reader = io::BufReader::new(f);

    let mut corruption = Vec::new();

    for line in reader.lines() {
        let line = line?;
        corruption.push(parse_pair(&line));
    }

    let result = try_find_path(1024, &corruption, edge, &start, &end).unwrap();
    println!("Part 1: {}", result);

    Ok(())
}

fn parse_pair(line: &str) -> (i32, i32) {
    let coords: Vec<&str> = line.split(",").map(|s| s.trim()).collect();
    let (x, y) = (coords[0].parse::<i32>().unwrap(), coords[1].parse::<i32>().unwrap());
    return (x, y)
}

fn try_find_path(n: usize, corruption: &Vec<(i32, i32)>, edge: i32, start: &(i32, i32), end: &(i32, i32)) -> Option<i32> {
    let corruption = &corruption[..n];
    let mut unvisited = HashSet::new();
    let mut dist = HashMap::new();
    dist.insert(*start, 0);
    for y in 0i32..=edge {
        for x in 0i32..=edge {
            if !corruption.contains(&(x, y)) {
                unvisited.insert((x, y));
            }
        }
    }
    while unvisited.len() > 0 {
        let current = get_closest(&unvisited, &dist);
        for considered in get_unvisited_neighbors(&current, &unvisited, &edge) {
            if !dist.contains_key(&current) {
                return None;
            }
            let new_cost = dist.get(&current).unwrap() + 1;
            if !dist.contains_key(&considered) || new_cost < *dist.get(&considered).unwrap() {
                dist.insert(considered, new_cost);
            }
        }
        unvisited.remove(&current);
        if dist.contains_key(&end) {
            break;
        }
    }
    Some(*dist.get(&end).unwrap())
}

fn get_closest(unvisited: &HashSet<(i32, i32)>, dist: &HashMap<(i32, i32), i32>) -> (i32, i32) {
    *unvisited.iter().min_by_key(|n| dist.get(n).unwrap_or(&i32::MAX)).unwrap()
}

fn get_unvisited_neighbors(p: &(i32, i32), unvisited: &HashSet<(i32, i32)>, edge: &i32) -> Vec<(i32, i32)> {
    let mut neighbors = Vec::new();
    for (dx, dy) in [(-1, 0), (1, 0), (0, -1), (0, 1)] {
        let n = (p.0 + dx, p.1 + dy);
        if n.0 >= 0 && n.0 <= *edge && n.1 >= 0 && n.1 <= *edge && unvisited.contains(&n) {
            neighbors.push(n);
        }
    }
    neighbors
}

